design/decisions/0008-extensible-config-management.md

```diff
@@ -0,0 +1,210 @@
+  after the config file update, as follows:
+  ```
+  [[files]]
+  type = "collectd"
+  path = "/etc/collectd/collectd.conf"
+  exec = "systemctl restart collectd"
+
+  [[files]]
+  type = "nginx"
+  path = "/etc/nginx/nginx.conf"
+  exec = "systemctl reload nginx"
+  ```
+- The `file` plugin's is implemented as follows:
+  - `list`: list all the config types listed in `tedge-configuration-plugin.toml`.
+  - `get`: Copy the contents of the existing configuration to the temp target file argument passed to it.
+  - `set`: Replace the existing config file with new config file in the argument.
+  - `finalize`: Perform the command specified in the `exec` field.
+    No-op when no `exec` command is provided.
+- For non file based configurations, that can't be defined in the `tedge-configuration-plugin.toml`,
+  they must have dedicated plugins that declare their types in the `list` command.
+
+The agent uses the plugins as follows:
+
+- On startup, gather the supported types from all the plugins by running the `list` command on them.
+  Those failing to execute the `list` command are not qualified as valid plugins and ignored.
+- Whenever the `/etc/tedge/config-plugins` directory is updated (new plugin installed or existing one removed),
+  agent refreshes the supported types.
+  Simply touching this directory would also trigger a refresh.
+- When the supported types are published over mqtt, their source plugin information is also appended to that type
+  in the format <config_type>::<plugin_type>.
+  For example, if a `mosquitto` plugin lists two different config types: `mosquitto.conf` and `mosquitto.acl`,
+  both types would be reported as `mosquitto.conf::mosquitto` and `mosquitto.acl::mosquitto`.
+- When a `config_snapshot` request is received for a type, call the `get` command of the corresponding plugin for that type
+  which can be derived from the `::` suffix of that type.
+  The `type` passed to the `get` command would not include the plugin suffix.
+  If there is no explicit suffix, it is delegated to the default file plugin.
+- When a `config_update` request is received for a type, the following actions are performed in sequence:
+  - Call `get` command of the corresponding plugin and cache the target temporary file as a backup.
+  - Call `set` command of the corresponding plugin with the updated config path in the argument.
+  - Call `finalize` command of the plugin and complete the operation if successful.
+  - If `set` or `finalize` commands fail, call the `set` command again with the original config file backup as the target.
+  - Call `finalize` again if the previous `set` attempt to restore teh original configuration succeeded and fail the operation.
+
+### Phase 2
+
+Instead of the `exeucting` phase of `config_update` workflow doing everything in that single step,
+break it down into multiple smaller stages that corresponds to each plugin command/phase as well as follows:
```

Comment:
I think we still need to evaluate the phase 2 proposal, as I'm not convinced that these builtin internal functions will be easy to use/document, but I believe with the PoC we might get some clarity in the right direction once we have something semi-functional.

Reply:
As an initial approach, I would even be tempted to make a `tedge` (or tedge-agent) command that would call the appropriate function with information provided as arguments, as this would be more re-usable for other integrations rather than relying on this workflow specific api calls.

Reply:
Config plugin spec will be refined further in a separate PR, although there was a general consensus on not going ahead with this specific phase 2 proposal.